File: typescript/src/workshop/m8/s04_shadowlimits/step2/ShadowBooking.ts
import type { BookingRequest } from '../BookingRequest.js';
import type { Infrastructure } from '../Infrastructure.js';
import { LegacyBookingFlow } from './LegacyBookingFlow.js';
import { NewBookingFlow } from './NewBookingFlow.js';
import { RecordingEffects } from './RecordingEffects.js';

/**
 * Krok 2: w cieniu kandydat dostaje RecordingEffects. Klient dostaje jeden mail i jedno
 * obciążenie, a cień porównuje nie tylko wynik, ale też ZAMIERZONE efekty z tymi, które
 * faktycznie wykonało legacy - i znajduje różnicę w treści maila.
 */
export class ShadowBooking {
  private readonly legacy: LegacyBookingFlow;
  private readonly divergenceList: string[] = [];

  constructor(private readonly infra: Infrastructure) {
    this.legacy = new LegacyBookingFlow(infra);
  }

  book(request: BookingRequest): string {
    const before = this.infra.log().length;
    const result = this.legacy.book(request);
    const legacyEffects = this.infra.log().slice(before);
    try {
      const recorder = new RecordingEffects();
      const shadow = new NewBookingFlow(recorder).book(request);
      this.compare('wynik', [result], [shadow]);
      this.compare('efekty', legacyEffects, recorder.recorded());
    } catch (failure) {
      this.divergenceList.push('kandydat: ' + (failure as Error).message);
    }
    return result;
  }

  private compare(what: string, legacyValues: readonly string[], candidateValues: readonly string[]): void {
    for (let i = 0; i < Math.max(legacyValues.length, candidateValues.length); i++) {
      const legacyValue = i < legacyValues.length ? legacyValues[i]! : '-';
      const candidateValue = i < candidateValues.length ? candidateValues[i]! : '-';
      if (legacyValue !== candidateValue) {
        this.divergenceList.push(what + ' legacy: ' + legacyValue + ' | kandydat: ' + candidateValue);
      }
    }
  }

  divergences(): readonly string[] {
    return Object.freeze([...this.divergenceList]);
  }
}

File: csharp/src/Training.Workshop/M8/S04ShadowLimits/Step2/ShadowBooking.cs
namespace Training.Workshop.M8.S04ShadowLimits.Step2;

/// <summary>
/// Krok 2: w cieniu kandydat dostaje RecordingEffects. Klient dostaje jeden mail i jedno
/// obciążenie, a cień porównuje nie tylko wynik, ale też ZAMIERZONE efekty z tymi, które
/// faktycznie wykonało legacy - i znajduje różnicę w treści maila.
/// </summary>
public sealed class ShadowBooking
{
    private readonly Infrastructure _infra;
    private readonly LegacyBookingFlow _legacy;
    private readonly List<string> _divergences = [];

    public ShadowBooking(Infrastructure infra)
    {
        _infra = infra;
        _legacy = new LegacyBookingFlow(infra);
    }

    public string Book(BookingRequest request)
    {
        int before = _infra.Log().Count;
        string result = _legacy.Book(request);
        IReadOnlyList<string> legacyEffects = _infra.Log().Skip(before).ToList();
        try
        {
            var recorder = new RecordingEffects();
            string shadow = new NewBookingFlow(recorder).Book(request);
            Compare("wynik", [result], [shadow]);
            Compare("efekty", legacyEffects, recorder.Recorded());
        }
        catch (Exception failure)
        {
            _divergences.Add("kandydat: " + failure.Message);
        }
        return result;
    }

    private void Compare(string what, IReadOnlyList<string> legacyValues, IReadOnlyList<string> candidateValues)
    {
        for (int i = 0; i < Math.Max(legacyValues.Count, candidateValues.Count); i++)
        {
            string legacyValue = i < legacyValues.Count ? legacyValues[i] : "-";
            string candidateValue = i < candidateValues.Count ? candidateValues[i] : "-";
            if (!legacyValue.Equals(candidateValue))
            {
                _divergences.Add(what + " legacy: " + legacyValue + " | kandydat: " + candidateValue);
            }
        }
    }

    public IReadOnlyList<string> Divergences()
    {
        return _divergences.ToList();
    }
}
